pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -255,6 +230,40 @@ public long getCommittedTxnCount() {
 
     @Override
     public CompletableFuture<Position> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
+        buffer.retain();
+        return transactionBufferFuture.thenCompose(ignore -> {
```

Comment:
Will it break the order of the request? 

If you requested 1,2,3
But after the transactionBufferFuture completed, it will execute as 3,2,1

It should be a problem, right?

Reply:
Thanks for pointing it out. This should be a problem.

Reply:
Thanks for pointing it out. This should be a problem.

Reply:
I introduce a new attribute `publishFuture` to resolve this problem. 

First, initialize `future` to `transactionBufferFuture.thenApply(__ -> PositionImpl.EARLIEST)`. Then, in each request, we add a new operation using the `thenCompose` method and update the result to a new `CompletableFuture`. In this way, each added operation will depend on the previous operation, ensuring that they are executed in order.

Reply:
@liangyepianzhou have you reproduce the issue before you push the commit for fixing this issue?